### PR TITLE
Add request-response and file downloading features

### PR DESCRIPTION
**Pull request by minudika**
1. Handling request-response scenarios.
2. File downloading.

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpResponseSource.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+                        description = "This parameter is used to map the http-response source to a " +
+                                "http-request sink. Then this source will accepts the response messages for " +
+                                "the requests sent by corresponding http-request sink.",
+                        type = {DataType.STRING}),
+                @Parameter(name = "http.status.code",
+                        description = "Acceptable http status code for the responses.\n" +
+                                "This can be a complete string or a regex.\n" +
+                                "Only the responses with matching status codes to the defined value, will be received" +
+                                " by the http-response source.\n" +
+                                "Eg: 'http.status.code = '200', http.status.code = '2*''",
```
Comment (suhothayan) [inline component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpResponseSource.java]: Fix 2** issue

File: component/src/main/java/org/wso2/extension/siddhi/io/http/sink/HttpRequestSink.java [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+                                "downloaded should be specified using 'download.path' parameter. This should be an " +
+                                "absolute path including the file name.",
+                        type = {DataType.STRING},
+                        optional = true,
+                        defaultValue = "null",
+                        dynamic = true),
+        },
+        examples = {
+                @Example(syntax =
+                        "@sink(type='http-request', sink.id='foo', publisher.url='http://localhost:8009/foo', " +
```
Comment (suhothayan) [inline component/src/main/java/org/wso2/extension/siddhi/io/http/sink/HttpRequestSink.java]: Should we define all these params? use the most use ones

File: component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpResponseSource.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+import static org.wso2.extension.siddhi.io.http.util.HttpConstants.DEFAULT_WORKER_COUNT;
+
+/**
+ * Http source for receive the http and https request.
+ */
+@Extension(
+        name = "http-response",
+        namespace = "source",
+        description = "" +
+                "The HTTP-RESPONSE co-relates with HTTP-REQUEST source with the parameter 'sink.id'.\n" +
```
Comment (suhothayan) [inline component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpResponseSource.java]: use consistent styles in the naming "HTTP-RESPONSE"

File: component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpResponseProcessor.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+                    logger.error("Error closing byte buffer.", e);
+                }
+            }
+        }
+    }
+
+    private String writeToTile(HTTPCarbonMessage carbonMessage) {
+        File file = new File(filePath);
+        try (
+                InputStream inputStream = new HttpMessageDataStreamer(carbonMessage).getInputStream();
```
Comment (suhothayan) [inline component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpResponseProcessor.java]: streams are not closed

File: component/src/main/java/org/wso2/extension/siddhi/io/http/sink/HttpSink.java [hunk 5]
```diff
@@ -767,6 +709,105 @@ private String getMessageBody(Object payload) {
...
+        }
+
+        String scheme = HttpSinkUtil.getScheme(publisherURL);
+        this.httpURLProperties = HttpSinkUtil.getURLProperties(publisherURL);
+        //Generate basic sender configurations
+        SenderConfiguration senderConfig = HttpSinkUtil
+                .getSenderConfigurations(httpURLProperties, clientStoreFile, clientStorePass, configReader);
+        if (EMPTY_STRING.equals(publisherURL)) {
+            throw new SiddhiAppCreationException("Receiver URL found empty but it is Mandatory field in " +
+                    "" + HttpConstants.HTTP_SINK_ID + "in" + streamID);
```
Comment (dilini-muthumala) [inline component/src/main/java/org/wso2/extension/siddhi/io/http/sink/HttpSink.java]: "in"  has to be surrounded with spaces.

File: component/src/main/java/org/wso2/extension/siddhi/io/http/sink/HttpSink.java [hunk 6]
```diff
@@ -767,6 +709,105 @@ private String getMessageBody(Object payload) {
...
+        if (HttpConstants.SCHEME_HTTPS.equals(scheme) && ((clientStoreFile == null) || (clientStorePass == null))) {
+            throw new ExceptionInInitializerError("Client trustStore file path or password are empty while " +
+                    "default scheme is 'https'. Please provide client " +
+                    "trustStore file path and password in" + streamID);
+        }
+        //if username and password both not equal to null consider as basic auth enabled if only one is null take it
+        // as exception
+        if ((EMPTY_STRING.equals(userName) ^
+                EMPTY_STRING.equals(userPassword))) {
+            throw new SiddhiAppCreationException("Please provide user name and password in " +
```
Comment (dilini-muthumala) [inline component/src/main/java/org/wso2/extension/siddhi/io/http/sink/HttpSink.java]: Log siddhi app name also when you throw SiddhiAppCreationException.

File: component/src/main/java/org/wso2/extension/siddhi/io/http/sink/HttpSink.java [hunk 7]
```diff
@@ -767,6 +709,105 @@ private String getMessageBody(Object payload) {
...
+                ProxyServerConfiguration proxyServerConfiguration = new ProxyServerConfiguration(proxyHost, Integer
+                        .parseInt(proxyPort));
+                if (!EMPTY_STRING.equals(proxyPassword) && !EMPTY_STRING.equals
+                        (proxyUsername)) {
+                    proxyServerConfiguration.setProxyPassword(proxyPassword);
+                    proxyServerConfiguration.setProxyUsername(proxyUsername);
+                }
+                senderConfig.setProxyServerConfiguration(proxyServerConfiguration);
+            } catch (UnknownHostException e) {
+                log.error("Proxy url and password is invalid in sink " + streamID, e);
```
Comment (dilini-muthumala) [inline component/src/main/java/org/wso2/extension/siddhi/io/http/sink/HttpSink.java]: Better log siddhi app name also here.

File: component/src/test/java/org/wso2/extension/siddhi/io/http/sink/util/HttpFileServerListener.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
+/*
+ *  Copyright (c) 2017 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (dilini-muthumala) [inline component/src/test/java/org/wso2/extension/siddhi/io/http/sink/util/HttpFileServerListener.java]: Fix year.

File: component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpResponseProcessor.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+                OutputStream outputStream = new FileOutputStream(file);
+
```
Comment (mohanvive) [inline component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpResponseProcessor.java]: Above code is not formatted properly. Please check.

File: component/src/main/java/org/wso2/extension/siddhi/io/http/util/ResponseSourceId.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
+package org.wso2.extension.siddhi.io.http.util;
```
Comment (mohanvive) [inline component/src/main/java/org/wso2/extension/siddhi/io/http/util/ResponseSourceId.java]: Add license header

File: component/src/test/java/org/wso2/extension/siddhi/io/http/sink/HttpRequestResponseTestCase.java [hunk 11]
```diff
@@ -0,0 +1,294 @@
+package org.wso2.extension.siddhi.io.http.sink;
```
Comment (mohanvive) [inline component/src/test/java/org/wso2/extension/siddhi/io/http/sink/HttpRequestResponseTestCase.java]: Add license header

File: component/src/test/java/org/wso2/extension/siddhi/io/http/sink/HttpSinkTestCase.java [hunk 12]
```diff
@@ -150,6 +158,59 @@ public void testHTTPContentTypeAtHeaders() throws Exception {
         siddhiAppRuntime.shutdown();
         lst.shutdown();
     }
-    
-}
 
+    //@Test
```
Comment (mohanvive) [inline component/src/test/java/org/wso2/extension/siddhi/io/http/sink/HttpSinkTestCase.java]: Why is annotation commented?

File: component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpResponseConnectorListener.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+        }
+        HttpResponseProcessor workerThread =
+                new HttpResponseProcessor(carbonMessage, sourceEventListener, sinkId, properties);
+        executorService.execute(workerThread);
+    }
+
+    @Override
+    public void onError(Throwable throwable) {
+        log.error("Error occurred during processing response for the request sent by http-request-sink with " +
+                "'sink.id' = " + sinkId, throwable);
```
Comment (grainier) [inline component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpResponseConnectorListener.java]: Better to include `siddhiAppName` as well.

File: component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpResponseProcessor.java [hunk 14]
```diff
@@ -0,0 +1,117 @@
+/*
+ *  Copyright (c) 2017 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (minudika) [reply component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpResponseProcessor.java]: Fixed with 8f6f25e287b11fbf2b21019bda6bbd8835b7e26f

File: component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpResponseProcessor.java [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+                    if (logger.isDebugEnabled()) {
+                        logger.debug("Empty payload event, hence dropping the event chunk in the source with sink" +
+                                ".id : " + sinkId);
+                    }
+                }
+            } finally {
+                try {
+                    buf.close();
+                } catch (IOException e) {
+                    logger.error("Error closing byte buffer.", e);
```
[the remark this answers is not in the document]
Comment (minudika) [reply component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpResponseProcessor.java]: Fixed with 8f6f25e

File: component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpResponseProcessor.java (hunk 9 above)
Comment (minudika) [reply component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpResponseProcessor.java]: Fixed with 8f6f25e

File: component/src/test/java/org/wso2/extension/siddhi/io/http/sink/HttpRequestResponseTestCase.java (hunk 11 above)
Comment (minudika) [reply component/src/test/java/org/wso2/extension/siddhi/io/http/sink/HttpRequestResponseTestCase.java]: Fixed with 8f6f25e

File: component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpResponseConnectorListener.java (hunk 13 above)
Comment (minudika) [reply component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpResponseConnectorListener.java]: Fixed with 8f6f25e

File: component/src/test/java/org/wso2/extension/siddhi/io/http/sink/util/HttpFileServerListener.java (hunk 8 above)
Comment (minudika) [reply component/src/test/java/org/wso2/extension/siddhi/io/http/sink/util/HttpFileServerListener.java]: Fixed with 8f6f25e

File: component/src/main/java/org/wso2/extension/siddhi/io/http/sink/HttpSink.java (hunk 7 above)
Comment (minudika) [reply component/src/main/java/org/wso2/extension/siddhi/io/http/sink/HttpSink.java]: Fixed with 8f6f25e

File: component/src/main/java/org/wso2/extension/siddhi/io/http/sink/HttpSink.java (hunk 5 above)
Comment (minudika) [reply component/src/main/java/org/wso2/extension/siddhi/io/http/sink/HttpSink.java]: Fixed with 8f6f25e

File: component/src/main/java/org/wso2/extension/siddhi/io/http/sink/HttpSink.java (hunk 6 above)
Comment (minudika) [reply component/src/main/java/org/wso2/extension/siddhi/io/http/sink/HttpSink.java]: Fixed with 8f6f25e

File: component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpResponseConnectorListener.java (hunk 13 above)
Comment (minudika) [reply component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpResponseConnectorListener.java]: Fixed with PR https://github.com/wso2-extensions/siddhi-map-json/pull/56

**Pull request merged**